### ISSUE TITLE
Docs: How to migrate from async router.push to new `next/navigation` push method

### ISSUE THREAD

**Issue by ajwootto** [vercel/next.js#61737]
### What is the improvement or update you wish to see?

The new router in `next/navigation` no longer returns a promise from the `push` method. From what I've been able to understand, this is so that it can work with the `startTransition` feature of React, but the Next docs don't contain any information about this or how to successfully replace any functionality that was relying on "awaiting" a router.push call

### Is there any context that might help us understand?

The docs for the push method are currently one sentence here:
https://nextjs.org/docs/app/api-reference/functions/use-router#userouter

The migration guide does not mention the fact that the method is no longer async or describe how a migration should take place to the new `push`
https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#step-5-migrating-routing-hooks

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#step-5-migrating-routing-hooks

**Comment by Guo-dalu** [comment]
I also need this docs badly, since given the current `useRouter`, it causes a lot of unexpected re-renders. Like, if I want to simply add some new querys in the url, I cannot wrap it in the useCallback --- it will only take the latest query rather than adding all the queries to the url.

```
export default function useRouterQuery() {
	const searchParams = useSearchParams()
	const pathname = usePathname()
	const router = useRouter()
	const replaceRouterQuery = useCallback(
		(queryKey: string, queryValue?: string, options?: Record<string, any>) => {
			const current = new URLSearchParams(searchParams)
			current.set(queryKey, queryValue || '')
			router.replace(`${pathname}?${current.toString()}`, { scroll: false, ...options })
		},
		[pathname, router, searchParams]
	)
	return replaceRouterQuery
}
```

**Comment by arcticfly** [comment]
Agreed. We built our own hacky solution to await router pushes, but would really like built-in support for this!

**Comment by ArianHamdi** [comment]
> Agreed. We built our own hacky solution to await router pushes, but would really like built-in support for this!

Can you please share your workaround

**Comment by arcticfly** [comment]
> > Agreed. We built our own hacky solution to await router pushes, but would really like built-in support for this!
> 
> Can you please share your workaround

// modify router so that it awaits the route change before returning
export const useAsyncRouter = () => {
  const router = useRouter();
  const pathname = usePathname();
  const [resolveRouteChanged, setResolveRouteChanged] = useState<(() => void) | null>(null);

  useEffect(() => {
    resolveRouteChanged?.();
  }, [pathname]);

  const asyncRouter = useMemo(() => {
    const push = async <RouteType>(href: NextRouteTypes.RouteImpl<RouteType>) => {
      router.push(href as any);

      await new Promise<void>((resolve) => {
        // wait for the route change to complete
        setResolveRouteChanged(resolve);
      });
    };

    return { ...router, push };
  }, [router]);

  return asyncRouter;
};

**Comment by Cauen** [comment]
Here are an alternative implementation of https://stackoverflow.com/a/77931487/6638583 workaround code

```ts
'use client'
import { useRouter } from 'next/navigation'
import { useEffect, useTransition } from 'react'

// Define the type for the observer callback function
type ObserverCallback = () => void

const createRouteObserver = () => {
  let observer: ObserverCallback | null = null

  const setObserver = (callback: ObserverCallback) => {
    observer = callback
  }

  const notify = () => {
    if (observer) {
      observer()
    }
  }

  return { setObserver, notify }
}

const routeObserver = createRouteObserver()

export const useAsyncRoutePush = () => {
  const [isPending, startTransition] = useTransition()
  const router = useRouter()

  const asynPush = async (path: string) => {
    return new Promise<void>((resolve) => {
      startTransition(() => {
        router.push(path)
      })

      routeObserver.setObserver(() => {
        resolve()
      })
    })
  }

  useEffect(() => {
    if (!isPending) {
      routeObserver.notify()
    }
  }, [isPending])

  return asynPush
}
```

Usage:

```tsx
export default function MyComponent() {
  const asyncPush = useAsyncRoutePush()

  return <Button onClick={async () => {
    await asyncPush('/')
    console.log("Route changed")
  }}>
    Go / route
  </Button>
}

```

**Comment by allicanseenow** [comment]
This should be mentioned in the documentation as a breaking change as there's no built-in support for this feature now.

**Comment by ivanbogdan-git** [comment]
+1

**Comment by HasanMothaffar** [comment]
+1 This feature is really needed

**Comment by scyllacz** [comment]
Still nothing? Is something what is keeping it back from implementing this in app router?

**Comment by lucastnr** [comment]
I would love it implemented too, it's a bummer that we have to do it manually

**Comment by daveidivide** [comment]
Same here, not enjoying writing hacks to get around the fact that router.push isn't async
